Neon Authorize docs pages (#2317)
neondatabase/website: content/docs/guides/neon-authorize.md

---
title: About Neon Authorize

> subtitle: Secure your application at the database level using Postgres's Role-Level Security

Author: Thanks for the catch!

> subtitle: Secure your application at the database level using Postgres's Role-Level Security

Reviewer: small thing, I omitted '-' from it but not sure that should still be in it, so "Row-Level Security". At least it should be consistent. I'll leave that to you 👍

enableTableOfContents: true
---

<InfoBlock>
<DocsList title="What you will learn:">
<p>JSON Web Tokens (JWT)</p>
<p>Row-level Security (RLS)</p>
<p>How Neon Authorize works</p>
</DocsList>

<DocsList title="Before you start" theme="docs">
<a href="/docs/guides/neon-authorize-quickstart">Get started</a>
</DocsList>

<DocsList title="Example repository" theme="repo">
<a href="https://github.com/orgs/neondatabase/repositories?type=source&q=authorize">6 sample demos</a>
</DocsList>
</InfoBlock>

<ComingSoon/>

**Neon Authorize** integrates with third-party **JWT-based authentication providers** like Auth0 and Clerk, bringing authorization closer to your data by leveraging **Row-Level Security (RLS)** at the database level.

## Authentication and authorization

> When implementing user authentication in your application, third-party authentication providers like **Clerk**, **Auth0**, and others simplify the process of managing user identities, passwords, and security tokens. Once a user's identity is confirmed, the next step is **authorization** — controlling who can do what in your app based on their user type or role — for example, admins versus regular users. With Neon Authorize, you can handle authorization right in Postgres, in addition to, or to replace entirely, security at other layers.

Reviewer 2: ```suggestion
When implementing user authentication in your application, third-party authentication providers like **Clerk**, **Auth0**, and others simplify the process of managing user identities, passwords, and security tokens. Once a user's identity is confirmed, the next step is **authorization** — controlling who can do what in your app based on their user type or role — for example, admins versus regular users. With Neon Authorize, you can manage authorization directly within Postgres, either alongside or as a complete replacement for security at other layers.
```

With **Neon Authorize**, the JWT is passed on to Neon, where you can make use of the validated user identity right in Postgres. To integrate with an authentication provider, add your provider's JWT discovery URL to your project. This lets Neon retrieve the necessary keys to validate the JWTs.

Behind the scenes, the [Neon Proxy](#the-role-of-the-neon-proxy) performs the validation, while the open source extension [pg_session_jwt](#how-the-pg_session_jwt-extension-works) makes the extracted `user_id` available to Postgres. You can then use **Row-Level Security (RLS)** policies in Postgres to enforce access control at the row level, ensuring that users can only access or modify data according to the defined rules. Since these rules are implemented directly in the database, they can offer a secure fallback — or even a primary solution — in case security in other layers of your application fail. See [when to rely on RLS](#when-to-rely-on-rls) for more information.

![neon authorize architecture](/docs/guides/neon_authorize_architecture.png)

> 

Reviewer 2: Should we add a note to mention that it's possible to use Neon Auth without an auth provider, as mentioned here?
https://neondb.slack.com/archives/C066CAC2NQK/p1729203665691569

Author: added

  await fetchWithDrizzle(async (db) => {
    return db.insert(schema.todos).values({
      task: newTodo.newTodo,
      isComplete: false,
      userId, // Explicitly ties todo to the user
    });
  });

  revalidatePath('/');
}
```

In this case, you have to:

- Check if the user is authenticated and their `userId` matches the data they are trying to modify.
- Handle both task creation and authorization in the backend code.

### After Neon Authorize (RLS in the database):

With Neon Authorize, you can let the database handle the authorization through **Row-Level Security** (RLS) policies:

```typescript
pgPolicy('create todos', {
  for: 'insert',
  to: 'authenticated',
  withCheck: sql`(select auth.user_id() = user_id)`,
});
```

Now, in your backend, you can simplify the logic, removing the user authentication checks and explicit authorization handling.

```typescript shouldWrap
export async function insertTodo(newTodo: { newTodo: string }) {
  await fetchWithDrizzle(async (db) => {
    return db.insert(schema.todos).values({
      task: newTodo.newTodo,
      isComplete: false,
    });
  });

  revalidatePath('/');
}
```

> This approach is flexible: you can manage RLS policies directly in SQL or work more declaratively using an ORM, applying authorization within your schema. This ability to handle both schema and authorization in one place can make it easier to maintain security.

Reviewer 2: Is it only Drizzle that will support declarative RLS at this point? If so, should we say " ...and ORM that supports RLS..."
And maybe this is for a separate guide, but we should link to the drizzle example somewhere when it's ready.
https://neondb.slack.com/archives/C066CAC2NQK/p1729169058382619

Author: Good point, I'll reword. 

And the Drizzle RLS sample: yes, that looks like maybe a separate guide or docs page. We need a deeper dive into the authenticated/anonymous roles and how you'd use them.

</InfoBlock>

<ComingSoon/>

**Neon Authorize** integrates with third-party **JWT-based authentication providers** like Auth0 and Clerk, bringing authorization closer to your data by leveraging [Row-Level Security (RLS)](https://www.postgresql.org/docs/current/ddl-rowsecurity.html) at the database level.

## Authentication and authorization

When implementing user authentication in your application, third-party authentication providers like **Clerk**, **Auth0**, and others simplify the process of managing user identities, passwords, and security tokens. Once a user's identity is confirmed, the next step is **authorization** — controlling who can do what in your app based on their user type or role — for example, admins versus regular users. With Neon Authorize, you can manage authorization directly within Postgres, either alongside or as a complete replacement for security at other layers.

## How Neon Authorize works

Most authentication providers issue **JSON Web Tokens (JWTs)** on user authentication to convey user identity and claims. The JWT is a secure way of proving that logged-in users are who they say they are &#8212; and passing that proof on to other entities.

> With **Neon Authorize**, the JWT is passed on to Neon, where you can make use of the validated user identity directly in Postgres. To integrate with an authentication provider, add your provider's JWT discovery URL to your project. This lets Neon retrieve the necessary keys to validate the JWTs.

Reviewer 2: ```suggestion
With **Neon Authorize**, the JWT is passed on to Neon, where you can make use of the validated user identity directly in Postgres. To integrate with an authentication provider, you will add your provider's JWT discovery URL to your Neon project. This lets Neon retrieve the necessary keys to validate the JWTs.
```

```typescript shouldWrap
import { neon } from '@neondatabase/serverless';

const sql = neon(process.env.DATABASE_AUTHENTICATED_URL, { authToken: myAuthProvider.getJWT() });

await sql(`select * from todos`);
```

> Behind the scenes, the [Neon Proxy](#the-role-of-the-neon-proxy) performs the validation, while the open source extension [pg_session_jwt](#how-the-pg_session_jwt-extension-works) makes the extracted `user_id` available to Postgres. You can then use **Row-Level Security (RLS)** policies in Postgres to enforce access control at the row level, ensuring that users can only access or modify data according to the defined rules. Since these rules are implemented directly in the database, they can offer a secure fallback — or even a primary solution — in case security in other layers of your application fail. See [when to rely on RLS](#when-to-rely-on-rls) for more information.

Reviewer 2: ```suggestion
Behind the scenes, the [Neon Proxy](#the-role-of-the-neon-proxy) performs the validation, while Neon's open source [pg_session_jwt](#how-the-pg_session_jwt-extension-works) extension makes the extracted `user_id` available to Postgres. You can then use **Row-Level Security (RLS)** policies in Postgres to enforce access control at the row level, ensuring that users can only access or modify data according to the defined rules. Since these rules are implemented directly in the database, they can offer a secure fallback — or even a primary authorization solution — in case security in other layers of your application fail. See [when to rely on RLS](#when-to-rely-on-rls) for more information.
```

```typescript shouldWrap
export async function insertTodo(newTodo: { newTodo: string }) {
  await fetchWithDrizzle(async (db) => {
    return db.insert(schema.todos).values({
      task: newTodo.newTodo,
      isComplete: false,
    });
  });

  revalidatePath('/');
}
```

This approach is flexible: you can manage RLS policies directly in SQL, or use an ORM to centralize them within your schema. Keeping both schema and authorization in one place can make it easier to maintain security. Some ORMs like Drizzle are adding support for declaritive RLS, which makes the logic easier to scan and scale.

## How Neon Authorize gets `auth.user_id()` from the JWT

Let's break down the RLS policy controlling who can **view todos** to see what Neon Authorize is actually doing:

<Tabs labels={["SQL", "Drizzle"]}>

<TabItem>

```sql
CREATE POLICY "view todos" ON "todos" AS PERMISSIVE
  FOR SELECT TO authenticated
  USING ((select auth.user_id() = user_id));
```

</TabItem>

<TabItem>

```typescript
pgPolicy('view todos', {
  for: 'select',
  to: 'authenticated',
  using: sql`(select auth.user_id() = user_id)`,
});
```

</TabItem>

</Tabs>

> This policy enforces that a user can only view their own `todos`. Here's how each component works together.

Reviewer 2: ```suggestion
This policy enforces that an authenticated user can only view their own `todos`. Here's how each component works together.
```

### What Neon does for you

> When your application makes a request, Neon validates the JWT by checking its signature and expiration date against the public keys. Once validated, Neon extracts the `user_id` from the JWT and uses it in the datasbase session, making it accessible for RLS.

Reviewer 2: ```suggestion
When your application makes a request, Neon validates the JWT by checking its signature and expiration date against the public keys. Once validated, Neon extracts the `user_id` from the JWT and uses it in the database session, making it accessible for RLS.
```

### How the `pg_session_jwt` extension works

The **pg_session_jwt** extension enables RLS policies to verify user identity directly within SQL queries:

```typescript
using: sql`(select auth.user_id() = user_id)`,
```

- `auth.user_id()`: This function, provided by `pg_session_jwt`, retrieves the authenticated user's ID from the JWT (it looks for it in the `sub` field).
- `user_id`: This refers to the `user_id` column in the `todos` table, representing the owner of each to-do item.

The RLS policy compares the `user_id` from the JWT with the `user_id` in the todos table. If they match, the user is allowed to view their own todos; if not, access is denied.

## When to rely on RLS

> For early-stage applications, **RLS** might offer all the security you need to scale your project. For more mature applications or architectures where multiple backends read from the same database, RLS centralizes authorization rules within the database itself. This way, every service that accesses your database can benefit from secure, consistent access controls without needing to reimplement them individually.

Reviewer 2: ```suggestion
For early-stage applications, **RLS** might offer all the security you need to scale your project. For more mature applications or architectures where multiple backends read from the same database, RLS centralizes authorization rules within the database itself. This way, every service that accesses your database can benefit from secure, consistent access controls without needing to reimplement them individually in each connecting application.
```

| **Amazon Cognito**     | ✅         | `https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/jwks.json`             | [docs](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html) |
| **Azure AD**           | ✅         | `https://login.microsoftonline.com/{tenantId}/discovery/v2.0/keys`                          | [docs](https://learn.microsoft.com/en-us/entra/identity-platform/access-tokens)                                               |
| **GCP Cloud Identity** | ✅         | `https://www.googleapis.com/oauth2/v3/certs`                                                | [docs](https://developers.google.com/identity/openid-connect/openid-connect#discovery)                                        |
| **Descope Auth**       | ✅         | `https://api.descope.com/{YOUR_DESCOPE_PROJECT_ID}/.well-known/jwks.json`                   | [docs](https://docs.descope.com/project-settings/jwt-templates)                                                               |

## Sample applications

> You can use these sample ToDo applications to get started using Neon Authorize with the following popular authentication providers.

Reviewer 2: ```suggestion
You can use these sample ToDo applications to get started using Neon Authorize with popular authentication providers.
```

<a href="https://github.com/neondatabase-labs/azure-ad-b2c-nextjs-neon-authorize" description="A Todo List built with Azure AD B2C, Next.js, and Neon Authorize (SQL from the Backend)" icon="github">Azure AD B2C + Neon Authorize</a>
<a href="https://github.com/neondatabase-labs/descope-react-frontend-neon-authorize" description="A Todo list built with Descope, Next.js, and Neon Authorize (SQL from the frontend)" icon="github">Descope + Neon Authorize</a>
<a href="https://github.com/neondatabase-labs/authorize-demo-custom-jwt" description="A demo of Neon Authorize with custom generated JWTs" icon="github">Neon Authorize with custom JWTs</a>
</DetailIconCards>

## Current limitations

While this feature is in its early-access phase, there are some limitations to be aware of:

- **Authentication provider requirements**: Your authentication provider must provider must support **Asymmetric Keys**. For example, **Supabase Auth** will not be compatible until asymetric key support is added. You can track progress on this item [here](https://github.com/orgs/supabase/discussions/29289).

> - **Connection type**: Your application must use **HTTP** to connect to Neon. At this time, **TCP** and **WebSockets** are not supported. This means you need to use the [Neon serverless driver](/docs/serverless/serverless-driver) over HTTP.

Reviewer 2: ```suggestion
- **Connection type**: Your application must use **HTTP** to connect to Neon. At this time, **TCP** and **WebSockets** connections are not supported. This means you need to use the [Neon serverless driver](/docs/serverless/serverless-driver) over HTTP.
```
